flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,420 @@
+				handleNewCheckpoint(barrierId);
+				handler.runInTaskThread(() -> handler.notifyCheckpoint(barrier, 0, 0), "notifyCheckpoint");
+			}
+
+			int channelIndex = handler.getFlattenedChannelIndex(channelInfo);
+			if (barrierId == currentCheckpointId && !barrierReceivedChannels[channelIndex]) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("{}: Received barrier from channel {} @ {}.", handler.taskName, channelIndex, barrierId);
+				}
+
+				barrierReceivedChannels[channelIndex] = true;
+
+				if (++numBarriersReceived == numOpenChannels) {
+					allBarriersReceivedFuture.complete(null);
+				}
+			}
+		}
+
+		@Override
+		public synchronized void notifyBufferReceived(Buffer buffer, InputChannelInfo channelInfo) {
+			if (!barrierReceivedChannels[handler.getFlattenedChannelIndex(channelInfo)]) {
+				channelStateWriter.addInputData(
+					currentCheckpointId,
+					channelInfo,
+					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+					buffer);
+			} else {
+				buffer.recycleBuffer();
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+			allBarriersReceivedFuture.completeExceptionally(new IllegalStateException("Closed CheckpointBarrierUnaligner"));
+		}
+
+		private synchronized void handleNewCheckpoint(long barrierId) {
+			if (!allBarriersReceivedFuture.isDone()) {
+				// we did not complete the current checkpoint, another started before
+				LOG.warn("{}: Received checkpoint barrier for checkpoint {} before completing current checkpoint {}. " +
+						"Skipping current checkpoint.",
+					handler.taskName,
+					barrierId,
+					currentCheckpointId);
+
+				// let the task know we are not completing this
+				long currentCheckpointId = this.currentCheckpointId;
+				handler.runInTaskThread(() ->
+					handler.notifyAbort(currentCheckpointId,
+						new CheckpointException(
+							"Barrier id: " + barrierId,
+							CheckpointFailureReason.CHECKPOINT_DECLINED_SUBSUMED)),
+					"notifyAbort");
+			}
+
+			currentCheckpointId = barrierId;
+			Arrays.fill(barrierReceivedChannels, false);
+			numBarriersReceived = 0;
+			allBarriersReceivedFuture = new CompletableFuture<>();
+			channelStateWriter.start(barrierId);
+		}
+
+		public synchronized void releaseBlocksAndResetBarriers() {
+			Arrays.fill(barrierReceivedChannels, true);
+			// the next barrier that comes must assume it is the first
+			numBarriersReceived = 0;
+		}
+
+		public synchronized CompletableFuture<Void> getAllBarriersReceivedFuture(long checkpointId) {
+			if (checkpointId < currentCheckpointId) {
+				return FutureUtils.completedVoidFuture();
+			}
+			if (checkpointId > currentCheckpointId) {
+				throw new IllegalStateException("Checkpoint " + checkpointId + " has not been started at all");
+			}
+			return this.allBarriersReceivedFuture;
+		}
+
+		public synchronized void processEndOfPartition() {
+			numOpenChannels--;
+		}
+
+		public void setCurrentCheckpointId(long currentCheckpointId) {
+			this.currentCheckpointId = currentCheckpointId;
+		}
```

Comment:
1. Should this method be `syncrhonized`?
2. Should it check that `currentCheckpointId` passed is greater than `this.currentCheckpointId`?

Reply:
1. Yes!
2. Yes, but probably `>=` since it may have already peeked a barrier with the same id.

Reply:
I see it throws an exception now:
```
checkState(currentReceivedCheckpointId >= this.currentReceivedCheckpointId);
```
But I actually thought it will do nothing if the condition is false.
I think this is possible:
1. receive barrier A, "enqueue" it
2. receive barrier B, handle it in `RemoteInputChannel`; remember it in `ThreadSafeUnaligner`
3. dequeue and process barrier A
On step 3 we'll get an exception.

WDYT?

Reply:
You are right and I even encountered it in a test run. I modified it to `this.currentReceivedCheckpointId = Math.max(currentReceivedCheckpointId, this.currentReceivedCheckpointId);`